### DISCUSSION TITLE
Serial device does not receive data

### DISCUSSION THREAD

**Discussion by HoWeBrz** [pymeasure/pymeasure#631] [Q&A] [answered]
I do have a serial device (FuseBoard) to wich I can write / read data when using the pure serial connection.
But when I create an PyMeasure-Instrument, I can write data, but cannot receive any data.
It seems the read-buffer does not get any data. Sometime it echoes back the sent data.

Here is the "functional" pure serial connection example:
```
import serial
import time
ser = serial.Serial()  			# get port handle
ser.port = 'COM8'				
ser.baudrate = 115200
ser.timeout = 1					#read timeout

ser.open()						# open port
ser.is_open						# true if open
ser.write(b"##JT:reset\n")
ser.write(b"##JD2:2\n")
ser.write(b"##JT:send\n")

time.sleep(5)
rxdat = str(ser.read(1000), 'utf-8')         #read 100bytes or timeout
print(rxdat)
```
and the "rxdat" looks like this:
```
##JT:reset
##JD2:2
##JT:send


JTAG Transmission... 
-------
##JR2:3


RxBuffer
---------
0x000: 00000003
0x001: 00000003
0x002: ffffffff
0x003: ffffffff
0x004: ffffffff
0x005: 00000003
0x006: ffffffff
0x007: ffffffff
0x008: ffffffff
0x009: 0000000f
0x00a: ffffffff
0x00b: ffffffff
0x00c: ffffffff
0x00d: 00001fff
0x00e: ffffffff
0x00f: ffffffff
```

This is the "non-read-functional" instrument implementation (write works, because I can toggle LEDs on the serial device):

```
from pymeasure.adapters.serial import SerialAdapter
from pymeasure.instruments import Instrument
import time

class FuseBoard(Instrument):
    """Represents a FuseBoard device"""

    def __init__(self, resource_name, baud_rate=115200, **kwargs):
        super(FuseBoard, self).__init__(
            resource_name,
            "FuseBoard",
            baud_rate=baud_rate,
            data_bits=8,
            stop_bits=10,
            parity=0,
            read_termination='\r',
            write_termination='\n',
            **kwargs
        )

#############################################
# main function startes here
if __name__ == "__main__":
    
    # init FuseBoard device
    #fb1 = FuseBoard("ASRL8::INSTR")
    fb1 = FuseBoard("COM8")
    
    # turn LEDs on/off with 1s pause
    led_on = "##DP011:11111111"
    led_off = "##DP011:00000000"
    fb1.write(led_on)
    time.sleep(1)
    fb1.write(led_off)
       
    # write JTAG sequence
    fb1.write("##JT:reset")    
    fb1.write("##JI:0")
    fb1.write("##JD2:2")
    fb1.write("##JT:send")

    time.sleep(5)
    rxdat = fb1.read()
    print(rxdat)
```
I played with the receive termination character, but no success yet.

Why does the instrument not receive any RX data - any idea? 
How can I debug this ?

**Comment by msmttchr** [comment]
Is the read hanging or it just returns nothing?
Possibly you need to add a timeout parameter and investigate with an interactive serial terminal what is the termination character.

From the pyserial log you have sent, it seems that the instrument echoes the commands received, but it looks like there's no termination character

**Reply by bilderbuchi** [reply]
👍 this looks like it hangs waiting for the termination character forever. Try adding a `timeout=1000` kwarg to your `__init__` implementation, then you should get something back. 
And yes, try to find out what the read termination character is and define it accordingly. 
If there is none, you should be able to use `fb1.adapter.connection.read_bytes(count)` to read the raw binary data.

**Reply by BenediktBurger** [reply]
There is a termination character: print creates new lines instead of printing "\n".

You defined the read_termination as "\r", but from the printing, it should be "\n" as well. Many instruments have symmetric termination charakters.

**Reply by HoWeBrz** [reply]
I tried with "timeout=1000" in __init__, but did not help.
I tried termination character "\n" and "\r" and "\n\r", etc.  without success.

When using the "pure" serial connection, the received data looks like this:
`##DR011:000000ff\r\r\n##DP011:00000000\r\n##JT:reset\r\n##JI:0\r\n##JD2:2\r\n##JD98:8022\r\n##JD100:1GPshield: ##DR011:00000000\r\r\n0062\r\n##JD109:160000\r\n##JT:send\r\n\r\r\n\r\r\nJTAG Transmission... \r\r\n-------\r\r\n##JP0:\r\r\n##JR2:0\r\r\n##JR98:fffffffffffffffff00000003\r\r\n##JR100:fffffffffffffffff00000003\r\r\n##JR109:ffffffffffffffffffff0000000f\r\r\n\r\r\n\r\r\nRxBuffer\r\r\n---------\r\r\n0x000: 00000000\r\r\n0x001: 00000003\r\r\n0x002: ffffffff\r\r\n0x003: ffffffff\r\r\n0x004: ffffffff\r\r\n0x005: 00000003\r\r\n0x006: ffffffff\r\r\n0x007: ffffffff\r\r\n0x008: ffffffff\r\r\n0x009: 0000000f\r\r\n0x00a: ffffffff\r\r\n0x00b: ffffffff\r\r\n0x00c: ffffffff\r\r\n0x00d: 00001fff\r\r\n0x00e: ffffffff\r\r\n0x00f: ffffffff\r\r\n`

It seems there is no dedicated termination character. I assume the "pure" connection times-out and whatever was received before the timout is in the RX buffer.

Now, using the command (including the timeout=1000)
```
rxdat1 = fb1.adapter.connection.read_bytes(1000)
print(rxdat1)
```
shows me the same result - this is working.

Question:
What is the difference between 

```
fb1.adapter.connection.read_bytes(1000) 
and
fb1.read() 
```

?

Why does the 2nd not work ?

**Reply by HoWeBrz** [reply]
It was working for some time, but now I get the following timeout error:
`VisaIOError: VI_ERROR_ASRL_OVERRUN (-1073807252): An overrun error occurred during transfer. A character was not read from the hardware before the next character arrived.`

I assume it is because I try to read more bytes as are available in the buffer and then the timout kicks in.

Do I need a dedicated "termination" sequence/characters, or can I just use the timeout to receive data ?

**Reply by BenediktBurger** [reply]
It is better to use the termination Charakter, as that is the expected way and you get the result immediately instead of waiting for the timeout.

From your raw communication, the read_termination should be "\r\n".

The read termination is the termination while reading(it reads until these are found) , write termination while sending (they are added while sending).

Read expects termination character (and returns strings). Read_bytes reads a certain number of bytes and returns a bytes object.

**Comment by HoWeBrz** [comment, marked as answer]
I consider this topic as answered, since I was able to receive data via serial.
I will open a new discussion for further questions.

Can I somehow mark this discussion as solved ?
